central/hash/manager/deduper.go

```diff
@@ -0,0 +1,249 @@
+package manager
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/alert"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/reflectutils"
+	"github.com/stackrox/rox/pkg/sensor/hash"
+	"github.com/stackrox/rox/pkg/stringutils"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+// Deduper is an interface to deduping logic used to determine whether an event should be processed
+type Deduper interface {
+	// GetSuccessfulHashes returns a map of key to hashes that were successfully processed by Central
+	// and thus can be persisted in the database as being processed
+	GetSuccessfulHashes() map[string]uint64
+	// ShouldProcess takes a message and determines if it should be processed
+	ShouldProcess(msg *central.MsgFromSensor) bool
+	// MarkSuccessful marks the message if necessary as being successfully processed, so it can be committed to the database
+	// It will promote the message from the received map to the successfully processed map
+	MarkSuccessful(msg *central.MsgFromSensor)
+	// ProcessSync processes the Sensor sync message and reconciles the successfully processed and received maps
+	ProcessSync()
+}
+
+// NewDeduper creates a new deduper from the passed existing hashes
+func NewDeduper(existingHashes map[string]uint64) Deduper {
+	existingEntries := make(map[string]*entry)
+	for k, v := range existingHashes {
+		existingEntries[k] = &entry{
+			val: v,
+		}
+	}
+	return &deduperImpl{
+		received:              make(map[string]*entry),
+		successfullyProcessed: existingEntries,
+		hasher:                hash.NewHasher(),
+	}
+}
+
+type entry struct {
+	val       uint64
+	processed bool
+}
+
+type deduperImpl struct {
+	hashLock sync.RWMutex
+	// received map contains messages that have been received but not successfully processed
+	received map[string]*entry
```

Comment:
If I'm understanding this correctly the `received` map (and the `.processed` property in entry) are used for reconciliation only. 

E.g.: when reconciling, check if `successfullyProcessed` entry is `.processed`. If not, check if we've received it in `received` map. If not, remove it. 

I wonder if you could've solved this with a single `map` but adding something like an Enum to `entry`? 

```go
const (
	notSeen    int = iota
	seen           = iota
	processed      = iota
	reconciled     = iota
)
```

Since this is being populated with all hash entries during startup, they should all get `notSeen` state. Once they are received: `seen`, when `MarkSuccessful` is called: `processed` and then finally after `SYNCE` message: `reconciled`.

Wouldn't this make it simpler and avoid needing to keep track of same values in two maps? Or am I missing something?

Reply:
I actually think in this case it's simpler with two maps. This is because if you consolidate the states into a single map, then you need to start filtering what you flush to the database. For example, I tried to consolidate it into one map, but when we are doing reconciliation then we need to mark the entries as not yet seen, but then you need to distinguish what you are flushing to the database separately otherwise you'll flush very little data because they will be marked as unseen